Applications/Spire/Include/Spire/Ui/ListModel.hpp

```diff
@@ -0,0 +1,183 @@
+        int m_index;
+      };
+
+      /** Indicates a value was removed from the model. */
+      struct RemoveOperation {
+
+        /** The index of the value removed. */
+        int m_index;
+      };
+
+      /** Indicates a value was moved from one index to another. */
+      struct MoveOperation {
+
+        /** The index of the value that was moved. */
+        int m_source;
+
+        /** The index that the value was moved to. */
+        int m_destination;
+      };
+
+      /** Indicates a value was updated. */
+      struct UpdateOperation {
+
+        /** The index of the updated value. */
+        int m_index;
+      };
+
+      /** Consolidates all basic operations. */
+      using Operation = typename boost::make_recursive_variant<AddOperation,
+        RemoveOperation, MoveOperation, UpdateOperation,
+        Details::ModelTransaction<boost::recursive_variant_>>::type;
+
+      /**
+       * An operation consisting of a list of sub-operations performed as
+       * single transaction.
+       */
+      using Transaction = typename boost::mpl::deref<
+        typename boost::mpl::advance<
+          typename boost::mpl::begin<Operation::types>::type,
+          boost::mpl::int_<
+            boost::mpl::size<Operation::types>::value - 1>>::type>::type;
+
+      /**
+       * Signals an operation was applied to this model.
+       * @param operation The operation that was applied.
+       */
+      using OperationSignal = Signal<void (const Operation&)>;
+
+      virtual ~ListModel() = default;
+
+      /** Returns the number of items in the model. */
+      virtual int get_size() const = 0;
+
+      /**
+       * Returns the value at a specified index.
+       * @throws <code>std::out_of_range</code> iff index is out of range.
+       */
+      virtual const std::any& at(int index) const = 0;
+
+      /**
+       * Returns the value at a specified index.
+       * @throws <code>std::out_of_range</code> iff index is out of range.
+       */
+      template<typename T>
+      const T& get(int index) const;
+
+      /**
+       * Sets the value at a specified index.
+       * @param index - The index to set.
+       * @param value - The value to set at the specified index.
+       * @return The state of the value at the <i>index</i>, or
+       *         <code>QValidator::State::Invalid</code> iff row or column is
+       *         out of range.
+       */
+      virtual QValidator::State set(int index, const std::any& value);
+
+      /** Connects a slot to the OperationSignal. */
+      virtual boost::signals2::connection connect_operation_signal(
+        const typename OperationSignal::slot_type& slot) const = 0;
+
+    protected:
+
+      /** Constructs an empty model. */
+      ListModel() = default;
+
+    private:
+      ListModel(const ListModel&) = delete;
+      ListModel& operator =(const ListModel&) = delete;
+  };
+
+  template<typename T>
+  const T& ListModel::get(int index) const {
```

Comment:
Why is there these two methods that have the same effect, but one is non-virtual?

Reply:
The at function returns std::any, whereas the get function can return the value of the specified type. The at method can be overridden by the child class to check the data boundary. The get function is the member method of the base class, and it can interpret std::any to the specified type. I think the advantage of the design is that the template function can't be spread to the derived classes.

Reply:
Oh I see, I didn't notice the return type difference.